src/bartiq/symbolics/ast_parser.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2024 PsiQuantum, Corp.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""AST-based expression parser.
+
+Contrary to pyparsing-based grammar from grammar.py, this module
+utilizes ast to parse an arithmetic expression as Python expresson.
+Since not all expressions that we use are correct Python expressions,
+the parser first preprocesses expressions and replaces unallowed
+syntax with carefully crafted function calls. In particular:
+
+1. Port designations are converted to Port function call, e.g.
+
+#in_0 -> Port(in_0)
+a.b.#out_1 -> a.b.Port(out_1)
+
+2. Wildcards are replaced by a call to wildcard function, e.g.
+
+~.a -> wildcard().a
+a~.b -> wildcard(a).b
+
+Also, the parser here needs substantially less functionallity from
+the "interpreter", i.e. something that constructs actual objects
+from parsed informations. This is because we actually construct at most
+binary expressions, and we assume that operators of the constructed
+objects behave like they should with built-in operators.
+"""
+import ast
+import operator
+import re
+from dataclasses import dataclass
+from functools import singledispatch, singledispatchmethod
+from typing import Callable, TypeVar
+from warnings import warn
+
+from .grammar import Interpreter
+from .sympy_interpreter import SympyInterpreter
+
+_BINARY_OP_MAP = {
+    ast.Mult: operator.mul,
+    ast.Add: operator.add,
+    ast.Div: operator.truediv,
+    ast.Sub: operator.sub,
+    ast.Mod: operator.mod,
+    ast.Pow: operator.pow,
+    ast.BitXor: operator.pow,
+    ast.FloorDiv: operator.floordiv,
+}
+
+
+_UNARY_OP_MAP = {ast.USub: operator.neg, ast.UAdd: lambda x: +x}
+
+
+TExpr = TypeVar("TExpr")
+
+_IDENTIFIER = r"[_a-zA-Z]\w*"
+_NAMESPACE_IDENTIFIER = rf"{_IDENTIFIER}(\.{_IDENTIFIER})*"
+_PORT_PATTERN = rf"#({_NAMESPACE_IDENTIFIER})"
+_WILDCARD_PATTERN = rf"(({_IDENTIFIER})?)~"
+
+
+@dataclass(frozen=True)
+class PreprocessingStage:
```

Comment:
[question] Shouldn't most of the classes & functions in this module be private?

Reply:
And most are! Thanks for noticing some were not private - I just changed it.